### PR TITLE
Support "optional" message fields in Protobuf API

### PR DESCRIPTION
**Pull request by timmoon10**
protobuf 3 does not support optional primitive fields. If a field is not set, it is interpreted as "zero" (`float`s are `0.0f`, `bool`s are `false`, etc.). This is inconvenient when you have a field where zero is a valid value and the default should be non-zero. For instance, I would like to add support to the embedding layer to ignore indices that are equal to a "pad index". Index 0 is a perfectly reasonable pad index (and probably a common one), so a good default would be something like -1.

Note that this only applies to primitive fields. There is still machinery to detect if a message field is set or not. Furthermore, protobuf already defines [wrapper messages for the primitive data types](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/wrappers.proto). This means we can specify an optional field in a `.proto` file like this:
```
  message Embedding {
    int64 num_embeddings = 1;
    int64 embedding_dim = 2;
    google.protobuf.Int64Value padding_idx = 3;
  }
```
These wrapper messages are only useful for singular fields, since they will always be set inside repeated fields. I agree that this is somewhat difficult to interpret. In a more ideal world, protobuf would support typedefs and we could rename it to something like `OptionalInt64`. A corresponding prototext file looks like:
```
# With pad index
layer {
  embedding {
    num_embeddings: 1000000
    embedding_dim: 50
    padding_idx { value: 0 }

  }
}

# Without pad index
layer {
  embedding {
    num_embeddings: 1000000
    embedding_dim: 50
  }
}
```

This PR adds machinery in the Python frontend so that optional fields are handled appropriately:
```python
# With pad index
y = lbann.Embedding(x, num_embeddings=1000, embedding_dim=50, padding_idx=0)

# Without pad index
z = lbann.Embedding(x, num_embeddings=1000, embedding_dim=50)
```

### CODE REVIEW

**Comment by ndryden** [comment]
We may want to subsequently adjust the way batchnorm statistics specifications are handled in the prototext, since that was a particularly ugly case of working around this.

**Comment by timmoon10** [comment]
[Bamboo passes on Pascal and Corona (I wasn't able to get an allocation on Lassen)](https://lc.llnl.gov/bamboo/browse/LBANN-TIM219-2)
